feat!: Introduce `Expression::Variadic` and `Coalesce` expressions

lbhm opened the pull request:
## What changes are proposed in this pull request?

This PR introduces `Expression::Variadic` as a new kind of expression that takes a list of expressions as input and performs an operation on them.

This change is necessary to support the `Coalesce` expression, which is also introduced in this PR. We will need a `Coalesce` to support the read path for row tracking since the row ID (and similarly, the default commit version) is calculated as `COALESCE(materialized_row_id, fresh_row_id)`.

### This PR affects the following public APIs

This PR extends the expression framework with `Expression::Variadic` and thus requires all implementing engines to adopt the new expression.

## How was this change tested?

New UT.

lbhm commented on kernel/src/engine/arrow_expression/evaluate_expression.rs:
```diff
@@ -290,3 +319,223 @@ pub fn evaluate_predicate(
+    }
+
+    // Cast arrays to common supertype and collect ArrayData for MutableArrayData
+    let (casted, array_data): (Vec<ArrayRef>, Vec<ArrayData>) = arrays
+        .iter()
+        .map(|arr| {
+            let casted_arr = if arr.data_type() == &supertype {
+                Ok(arr.clone())
+            } else {
+                cast(arr, &supertype)
+            };
+            casted_arr.map(|arr| {
+                let data = arr.to_data();
+                (arr, data)
+            })
+        })
+        .collect::<Result<Vec<_>, ArrowError>>()?
+        .into_iter()
+        .unzip();
+
+    // Build result
+    let mut mutable = MutableArrayData::new(array_data.iter().collect(), false, num_rows);
```
`MutableArrayData::new()` requires a reference to owned `ArrayData`. I couldn't convince the compiler to do this in a more succinct (idiomatic?) way.
LMK if you see a better way.

scovich commented on kernel/src/expressions/mod.rs:
```diff
@@ -219,6 +226,14 @@ pub struct BinaryExpression {
+    /// The operator.
+    pub op: VariadicExpressionOp,
+    /// The input expressions.
+    pub exprs: Vec<Expression>,
```
aside: As part of the broader performance story, we should probably file a tracking issue to make all of these expression types use `Vec<ExpressionRef>`, basically finishing what @zachschuermann started with
* https://github.com/delta-io/delta-kernel-rs/pull/1171

(doesn't need to be this PR tho!)

lbhm replied:
I created https://github.com/delta-io/delta-kernel-rs/issues/1215 to track this.

scovich commented on kernel/src/expressions/transforms.rs:
```diff
@@ -279,6 +292,17 @@ pub trait ExpressionTransform<'a> {
+        &mut self,
+        v: &'a VariadicExpression,
+    ) -> Option<Cow<'a, VariadicExpression>> {
+        let nested_result = recurse_into_children(&v.exprs, |e| self.transform_expr(e))?;
+        Some(nested_result.map_owned_or_else(v, |exprs| VariadicExpression::new(v.op, exprs)))
```
What should happen if the transform removed all children?
* AND/OR - empty operator is well-defined (produces TRUE/FALSE)
* STRUCT - the operator itself gets removed

I lean toward the latter, since it's easy to argue that an empty `COALESCE()` should produce `NULL`: 
* The operator is defined to return the first non-NULL argument, or NULL if no such argument exists. 
* No arguments => no non-NULL arguments => return NULL

But we need to make sure we actually document and implement it that way (e.g. in default expression evaluator), if that's the decision.

scovich reviewed (approved):
Approach looks good. I guess adding support for COALESCE to the default engine's expression eval is a fast-follow? Otherwise it's not very useful...

lbhm commented on kernel/src/engine/arrow_expression/evaluate_expression.rs:
```diff
@@ -290,3 +319,223 @@ pub fn evaluate_predicate(
+///
+/// For each row, picks the first non-null value across the arrays. If the arrays are
+/// of different types, they must be castable to a common supertype.
+/// If `result_type` is provided, it must match the common supertype derived by [`common_supertype`].
```
I overhauled the type promotion function and only perform lossless type promotions now.

scovich replied:
More likely, the engine already has its own COALESCE implementation for the query path, and won't be using kernel expressions nor default evaluation at all (e.g. delta-rs, duckdb).

scovich commented:
> @scovich I don't have a strong opinion on type promotion. If you and the others think that we don't need it, I'm fine with removing it. I mainly thought it would be nice to have, but if there is no foreseeable use case, I agree with you.

Another way to look at it -- this PR could have merged a couple of days ago, except for the complexity and controversy of type promotion. For that reason, I would suggest moving the type promotion part to a different PR, to unblock this one.

lbhm commented on kernel/src/engine/arrow_expression/evaluate_expression.rs:
```diff
@@ -411,6 +443,255 @@ pub fn evaluate_predicate(
+        (Float32, Int8 | Int16 | UInt8 | UInt16) => Some(Float32),
+        (Int8 | Int16 | UInt8 | UInt16, Float32) => Some(Float32),
+        (Float64, Int8 | Int16 | Int32 | UInt8 | UInt16 | UInt32) => Some(Float64),
+        (Int8 | Int16 | Int32 | UInt8 | UInt16 | UInt32, Float64) => Some(Float64),
```
Removed type promotion

lbhm commented on kernel/src/engine/arrow_expression/evaluate_expression.rs:
```diff
@@ -558,57 +553,43 @@ pub fn coalesce_arrays(
-        types.push(arr.data_type());
+        if arr.data_type() != expected_type {
+            return Err(ArrowError::InvalidArgumentError(format!(
+                "Array at index {i} has type {:?}, but expected {:?}",
```
Ooops, thank you for flagging this. I overlooked it after adding `skip(1)`.

lbhm commented on kernel/src/engine/arrow_expression/evaluate_expression.rs:
```diff
@@ -411,15 +443,175 @@ pub fn evaluate_predicate(
+    arrays: &[ArrayRef],
+    result_type: Option<&DataType>,
+) -> Result<ArrayRef, ArrowError> {
+    if arrays.is_empty() {
+        return Err(ArrowError::InvalidArgumentError(
+            "The default engine currently does not support empty COALESCE statements".into(),
+        ));
+    }
+
+    // Early exit for single array case
+    if arrays.len() == 1 {
+        let array = &arrays[0];
+
+        // Validate result type if provided
+        if let Some(result_type) = result_type {
+            let result_arrow_type: ArrowDataType = result_type.try_into_arrow()?;
+            if array.data_type() != &result_arrow_type {
+                return Err(ArrowError::InvalidArgumentError(format!(
+                    "Single array type {:?} does not match requested result type {result_type:?}",
+                    array.data_type()
+                )));
+            }
+        }
+
+        return Ok(array.clone());
+    }
+
+    // Validate all arrays have the same length and same data type
+    let num_rows = arrays[0].len();
+    let expected_type = arrays[0].data_type();
+    for (i, arr) in arrays.iter().skip(1).enumerate() {
+        if arr.len() != num_rows {
```
Very elegant! I changed it.

zachschuermann commented on kernel/src/engine/arrow_expression/evaluate_expression.rs:
```diff
@@ -477,15 +492,97 @@ pub fn to_json(input: &dyn Datum) -> Result<ArrayRef, ArrowError> {
+/// - **Mismatched row counts**: Not all arrays have the same number of rows.
+/// - **Mismatched data types**: Not all arrays have exactly the same data type.
+/// - **Invalid result type**: If `result_type` is provided but doesn't match the arrays' data type.
+pub fn coalesce_arrays(
```
does it need to be `pub`? it's really just powering `COALESCE` inside expression eval right?

lbhm replied:
If we wanted to benchmark it in the expression_bench at a later point, it would need to be `pub`, right?

I can make `coalesce_arrays()` private for now since we don't use it in benchmarks yet, but `to_json()` (same context, different expression) already is `pub` and used outside the crate in our expression benchmark.

zachschuermann replied:
I don't think its worth blocking on but I'd prefer to keep pub(crate) or make `#[internal_api]` if we need? can make a follow up to normalize decision with to_json side as well

lbhm replied:
Agreed. I created https://github.com/delta-io/delta-kernel-rs/issues/1316 to track this.

zachschuermann commented on kernel/src/expressions/transforms.rs:
```diff
@@ -118,6 +118,15 @@ pub trait ExpressionTransform<'a> {
 
+    /// Called for each [`VariadicExpression`] encountered during the traversal. Implementations can
+    /// call [`Self::recurse_into_expr_variadic`] if they wish to recursively transform the children.
+    fn transform_expr_variadic(
```
are the expression transform changes sufficiently tested?

lbhm replied:
I added tests for `transform_expr_variadic`.

zachschuermann reviewed (approved):
LGTM, good to merge after:
1. let's make sure to call out this is a breaking change (`feat!`) and in PR desc - we are adding `Expression::Variadic` variant to a non-exhaustive enum.
2. check on tests for expression transform changes (need to add?)
3. consider making `coalesce_arrays` private

lbhm commented:
Thank you @zachschuermann.

- I updated the PR title and description.
- I added tests for `transform_expr_variadic`.
- I responded to your comment about making `coalesce_arrays` private. I don't have a strong opinion here, but wanted to raise one more aspect to consider.
